### ISSUE TITLE
3.0 update: Custom model binder

### ISSUE THREAD

**Issue by Rick-Anderson** [aspnet/AspNetCore.Docs#13529]
Remove images

---
#### Document Details

⚠ *Do not edit this section. It is required for docs.microsoft.com ➟ GitHub issue linking.*

* ID: cfb00531-4ac6-8bf6-9bd8-1eed6446397f
* Version Independent ID: ed418d09-6c33-699a-0d48-9aabc99457be
* Content: [Custom Model Binding in ASP.NET Core](https://docs.microsoft.com/en-us/aspnet/core/mvc/advanced/custom-model-binding?view=aspnetcore-2.2)
* Content Source: [aspnetcore/mvc/advanced/custom-model-binding.md](https://github.com/aspnet/AspNetCore.Docs/blob/master/aspnetcore/mvc/advanced/custom-model-binding.md)
* Product: **aspnet-core**
* Technology: **aspnetcore-mvc**
* GitHub Login: @ardalis
* Microsoft Alias: **riande**

**Comment by serpent5** [comment]
Do you want this and the 2.2 sample to switch over to snapshot files or do you want full samples for both?

RE the topic itself:

I think the explanation of the built-in `ByteArrayModelBinder` could be dropped in favour of an explanation of the built-in `CancellationTokenModelBinder`. This model-binder is a bit simpler and avoids raising the security issue re filenames, which is  already covered in the topic specific to file uploads.

This topic could still also mention `ByteArrayModelBinder` but without the sample code and associated security warnings. What do you think?

For context, `CancellationTokenModelBinder` kicks in if you have something like this:

```cs
public async Task<IActionResult> GetAuthors(CancellationToken cancellationToken)
```

I've also got a question about the note for the `AuthorEntityBinder` implementation, which states:

> For lookup, bind the `authorId` and query the database in an action method. This approach separates model binding failures from `NotFound` cases.

This separation appears to be implemented further down in `GetById`. The custom model-binder is still used but a simple `null` check handles the 404 and a `ModelState` check handles the 400. Do you think the note in question is still relevant?

**Comment by Rick-Anderson** [comment]
We're only keeping the 2.2 and 3.0 samples. Full samples are better. 

>I think the explanation of the built-in `ByteArrayModelBinder` could be dropped in favour of an explanation of the built-in `CancellationTokenModelBinder`. This model-binder is a bit simpler and avoids raising the security issue re filenames, which is  already covered in the topic specific to file uploads.

>This topic could still also mention `ByteArrayModelBinder` but without the sample code and associated security warnings. What do you think?

Agreed on both  points

>Do you think the note in question is still relevant?

Feel free to remove it.

Do you think the samples in this topic are reasonable? Seems like we could come up with simpler samples on when you'd need custom model binding.

**Comment by serpent5** [comment]
> Do you think the samples in this topic are reasonable? Seems like we could come up with simpler samples on when you'd need custom model binding.

I feel like the EF Core sample could be replaced with something similar, but I don't have any ideas for what to replace it with. On the plus side, the current sample does work well for covering general model-binding plus the extras like using DI. Do you have a suggestion for an alternative?

**Comment by Rick-Anderson** [comment]
@stevejgordon @dougbu  can you suggest a better sample?
[Customizing ASP.​NET Core Part 08: ModelBinders](https://asp.net-hacker.rocks/2018/10/17/customizing-aspnetcore-08-modelbinders.html) looks like a better sample. You can't copy that example, but you could do something very similar.

**Comment by serpent5** [comment]
Whenever I think of something that could be a custom model-binder, it seems there's a better option that's already built-in. Perhaps I'm mistaken, but for the example linked above I'd expect that to be implemented as an `InputFormatter`.

**Comment by Rick-Anderson** [comment]
> Whenever I think of something that could be a custom model-binder, it seems there's a better option that's already built-in. Perhaps I'm mistaken, but for the example linked above I'd expect that to be implemented as an `InputFormatter`.

If @stevejgordon @dougbu can't think of a better example, it's OK to use one that could be done with  `InputFormatter` and state that. 

I'd like to remove EF from the sample if possible - just to keep it simpler.

**Comment by dougbu** [comment]
@pranavkm any suggestions? (Sorry, I don't have time this week to refresh my memory of this sample.)

**Comment by serpent5** [comment]
Jon Skeet [answered](https://stackoverflow.com/a/57754748/2630078) a question on Stack Overflow today re using a custom model-binder with a type from his NodaTime library. Would it be acceptable to use this as the basis for the docs update here? Jon’s implementation is simplified, so the docs example could be more complete and demonstrate more. I guess my question is can we use a NodaTime `Instant` model-binder for the example? I dunno if it’s acceptable to pull in NodaTime for the example project, but the need for a custom model-binder here is a practical example of its usefulness.

**Comment by Rick-Anderson** [comment]
@serpent5  does that model bind a NodaTime string into a NodaTime type? I don't want to introduce a 3rd party library and all the complexity for a model binding sample. How about UTC time?

@pranavkm  can you think of a custom model binding sample for doc purposes?

**Comment by serpent5** [comment]
Yeah, it would introduce a third-party package. That's why I asked and your response is what I thought it might be. Fair enough.

**Comment by pranavkm** [comment]
@Rick-Anderson could you summarize what the goal of this issue is? Is it to produce a simple example for a custom model binder?

**Comment by Rick-Anderson** [comment]
@pranavkm the goal is that customers that need to create a custom model binder have a simple example. The sample code doesn't have to be practical/useful, it just needs to show how to do custom model binding. The sample I like best is **[this one](https://www.dotnetcurry.com/aspnet-mvc/1368/aspnet-core-mvc-custom-model-binding)**. Search for ` demonstrate custom model binders`

This is the scenario where `dateTime` fields are provided as separated date and time properties, but you still want to bind them to a single `dateTime` field in your model.

Important note: I like this as an example for model binders because it is easy enough to understand and has already been used as an example in the past. However, I wouldn’t use this on a real system where I would just stick with timestamps or ISO Strings and use JavaScript to combine values from multiple editors.

Let’s say we have an appointment model like this:

```
public class Appointment{    
    public string Id { get; set; }   
   public string Name { get; set; }    
  public DateTime AppointmentDate { get; set; }
}
```

Does that seem like a reasonable  approach to replace the exiting sample which is too complex and has security problems.

**Comment by pranavkm** [comment]
> Does that seem like a reasonable approach to replace the exiting sample which is too complex and has security problems.

Yeah that sounds fine. I'd use the first sample (without the fallback) to illustrate how to write a custom binder (the fallback while nice might adds a lot of complexity).

**Comment by Rick-Anderson** [comment]
@serpent5  should I take a stab at this and you can review/help if I get stuck? Feel free to do the PR if you'd rather do it. In that case, It might be easier if I start the PR and you take over my branch,

**Comment by serpent5** [comment]
Go ahead. I'll do whatever you need me to.

**Comment by Rick-Anderson** [comment]
@serpent5  Let me know if I should re-open this.
